### PR TITLE
Add shuffling function

### PR DESCRIPTION
**Pull request by yx1215**
*Issue #, if available:*

*Description of changes:*

*Note: replaces #858*

Added ShuffleIter class to enable shuffling  in data loader.

In the first time comparison plot, the x-axis represents the shuffle buffer length, that is, if the x-axis is c, then the shuffle buffer is of size c * batch size, y-axis is the time used to generate 200 batches of size 32 from a dataset of length 300.

comparison

In the following heatmaps, x-axis is the batch index, y-axis is the time series index, and color intensity indicates how many elements in the x-batch are from the y-time-series

single_process_1
single_process_10
two_process_1
two_process_10

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by AaronSpieler** [comment]
If I read you code correctly, you are shuffling the batches, but not the samples from which then different batches are generated?

**Comment by yx1215** [comment]
> If I read you code correctly, you are shuffling the batches, but not the samples from which then different batches are generated?

I am shuffling the samples which are put into each batch. Originally samples are generated sequentially from the function _sequential_sample_generator. Now I take these samples and put them in a shuffle buffer(a list) which has length larger than batch size, randomly generate a index, take the element at that index, put it in the batch and replace it with the next element in the original sequential generator. Also you can refer to the heat map in the PR. The x-axis means the batch number, and the y-axis means the time series id. So if the number at (x, y) is z, that means in batch number x, it takes z element from time series number y. It clearly shows that with a larger buffer size, the elements will become more shuffled.

**Comment by AaronSpieler** [comment]
> > If I read you code correctly, you are shuffling the batches, but not the samples from which then different batches are generated?
> 
> I am shuffling the samples which are put into each batch. Originally samples are generated sequentially from the function _sequential_sample_generator. Now I take these samples and put them in a shuffle buffer(a list) which has length larger than batch size, randomly generate a index, take the element at that index, put it in the batch and replace it with the next element in the original sequential generator. Also you can refer to the heat map in the PR. The x-axis means the batch number, and the y-axis means the time series id. So if the number at (x, y) is z, that means in batch number x, it takes z element from time series number y. It clearly shows that with a larger buffer size, the elements will become more shuffled.

Ok, got it, thx!

**Comment by AaronSpieler** [comment]
Ok, looks nice. Could you add some comments to the `ShuffleIter` class maybe?
Also, how did you test the speed of the respective shuffle setting?

**Comment by richardk53** [comment]
It looks like it will have a semi-random stopping time, I think between dataset_length - shuffle_buffer_length and dataset_length iterations (where dataset_length is the number of iterations until the base_iterator is exhausted).

E.g. if shuffle_buffer_length has the same length as the dataset, then the base_iterator will be exhausted in the first iteration. It will start putting Nones in the list until ShuffleIter randomly encounters one. This happens latest after all elements have been replaced with None.

**Comment by yx1215** [comment]
> Ok, looks nice. Could you add some comments to the `ShuffleIter` class maybe?
> Also, how did you test the speed of the respective shuffle setting?

I've updated the description in the PR about the plots. And comments are added to the shuffle iter.

**Comment by lostella** [comment]
> E.g. if shuffle_buffer_length has the same length as the dataset, then the base_iterator will be exhausted in the first iteration. It will start putting Nones in the list until ShuffleIter randomly encounters one. This happens latest after all elements have been replaced with None.

I think this may not be a problem, since at training time the data loader is initialized as "cyclic", see here: https://github.com/awslabs/gluon-ts/blob/master/src/gluonts/dataset/loader.py#L166. In fact, I think the same mechanism is used to ensure that if in an epoch you want to iterate more data than the time series you have (by setting num_batches_per_epoch) then the data loader starts back from the beginning once the dataset is exhausted.

The logic is a bit convoluted probably, and we should improve it.

**Comment by lostella** [comment]
Maybe related to what @richardk53 mentioned: the following snippet gives weird results

```python
data = range(20)
shuffled_data = ShuffleIter(base_iterator=iter(data), shuffle_buffer_length=10)
print(len(list(shuffled_data)))  # I would expect 20 here
```

I would expect to the shuffled data to always have the same size as the original data, in case that's finite (so 20 in the above example). But I'm getting all sorts of numbers instead, and sometimes errors.

@yx1215 I think `ShuffleIter` as a standalone component should should have that property right? Maybe you can fix it to make sure it does, and add a test for it?

**Comment by yx1215** [comment]
> Maybe related to what @richardk53 mentioned: the following snippet gives weird results
> 
> ```python
> data = range(20)
> shuffled_data = ShuffleIter(base_iterator=iter(data), shuffle_buffer_length=10)
> print(len(list(shuffled_data)))  # I would expect 20 here
> ```
> 
> I would expect to the shuffled data to always have the same size as the original data, in case that's finite (so 20 in the above example). But I'm getting all sorts of numbers instead, and sometimes errors.
> 
> @yx1215 I think `ShuffleIter` as a standalone component should should have that property right? Maybe you can fix it to make sure it does, and add a test for it?

**Comment by yx1215** [comment]
> Maybe related to what @richardk53 mentioned: the following snippet gives weird results
> 
> ```python
> data = range(20)
> shuffled_data = ShuffleIter(base_iterator=iter(data), shuffle_buffer_length=10)
> print(len(list(shuffled_data)))  # I would expect 20 here
> ```
> 
> I would expect to the shuffled data to always have the same size as the original data, in case that's finite (so 20 in the above example). But I'm getting all sorts of numbers instead, and sometimes errors.
> 
> @yx1215 I think `ShuffleIter` as a standalone component should should have that property right? Maybe you can fix it to make sure it does, and add a test for it?

I've fixed the problem in the following way: once the base iterator reaches its end, each time we select an element from the shuffle buffer, we will delete that element instead of replacing it. This, I think, will not have much influence on the performance because this is only used to deal with edge case. And I've updated the time comparison plot.

**Comment by richardk53** [comment]
> > E.g. if shuffle_buffer_length has the same length as the dataset, then the base_iterator will be exhausted in the first iteration. It will start putting Nones in the list until ShuffleIter randomly encounters one. This happens latest after all elements have been replaced with None.
> 
> I think this may not be a problem, since at training time the data loader is initialized as "cyclic", see here: https://github.com/awslabs/gluon-ts/blob/master/src/gluonts/dataset/loader.py#L166. In fact, I think the same mechanism is used to ensure that if in an epoch you want to iterate more data than the time series you have (by setting num_batches_per_epoch) then the data loader starts back from the beginning once the dataset is exhausted.
> 
> The logic is a bit convoluted probably, and we should improve it.


I think this component should not need to make these assumptions about other components using it. 
With the fix it should be possible to implement e.g. sampling without replacement (which often works better) by choosing the buffer size as the dataset length and iterating only once over the dataset.

**Comment by lostella** [comment]
> I think this component should not need to make these assumptions about other components using it.

Yes, totally agree. @yx1215 has fixed some of these edge cases already, I’ll check if that covers them all (we should test it well)

File: src/gluonts/dataset/parallelized_loader.py [hunk 1]
```diff
@@ -346,6 +357,51 @@ def _worker_fn(
...
+        if not self.shuffle_buffer:
+            self.shuffle_buffer = list(
+                itertools.islice(
+                    self.base_iterator, self.shuffle_buffer_length
+                )
+            )
+        # if buffer still empty, means all elements used,
+        # return a signal of end of iterator
+        if not self.shuffle_buffer:
+            return {}
```
Comment (lostella) [inline src/gluonts/dataset/parallelized_loader.py:381]: I think the right signal to send when the iterator is exhausted is raising a `StopIteration` exception, see here: https://docs.python.org/3/library/stdtypes.html#iterator-types

File: src/gluonts/dataset/parallelized_loader.py [hunk 2]
```diff
@@ -346,6 +357,51 @@ def _worker_fn(
...
+            del self.shuffle_buffer[idx]
+
+        return next_sample
+
+    def __iter__(self) -> Iterator[DataEntry]:
+        while True:
+            next_sample = next(self)
+            if not next_sample:
+                return
+            yield next_sample
```
Comment (lostella) [inline src/gluonts/dataset/parallelized_loader.py:402]: I think this can simply return `self`, see here: https://docs.python.org/3/library/stdtypes.html#iterator-types

File: src/gluonts/dataset/parallelized_loader.py (hunk 1 above)
Comment (yx1215) [reply src/gluonts/dataset/parallelized_loader.py:381]: yes, already changed.

File: src/gluonts/dataset/parallelized_loader.py (hunk 2 above)
Comment (yx1215) [reply src/gluonts/dataset/parallelized_loader.py:402]: Actually I found out that we don't need to implement ```__iter__``` method. Changing ```__next__``` is everything needed.

**Comment by AaronSpieler** [comment]
So it seems there is in general some documentation missing regarding `shuffle_buffer_length` in `loader.py` and  in `parallelized_loader.py`.

Additionally, I guess in most cases, the argument shuffle_buffer_length is implicitly passed as an argument. It should, however , be explicit just like `num_workers` or `num_prefetch`. Example below:

```
    def train(
        self,
        training_data: Dataset,
        validation_data: Optional[Dataset] = None,
        num_workers: Optional[int] = None,
        num_prefetch: Optional[int] = None,
        **kwargs,
    ) -> Predictor:
        has_negative_data = any(np.any(d["target"] < 0) for d in training_data)
        low = -10.0 if has_negative_data else 0
        high = 10.0
        bin_centers = np.linspace(low, high, self.num_bins)
        bin_edges = np.concatenate(
            [[-1e20], (bin_centers[1:] + bin_centers[:-1]) / 2.0, [1e20]]
        )

        logging.info(
            f"using training windows of length = {self.train_window_length}"
        )

        transformation = self.create_transformation(
            bin_edges, pred_length=self.train_window_length
        )

        training_data_loader = TrainDataLoader(
            dataset=training_data,
            transform=transformation,
            batch_size=self.trainer.batch_size,
            num_batches_per_epoch=self.trainer.num_batches_per_epoch,
            ctx=self.trainer.ctx,
            num_workers=num_workers,
            num_prefetch=num_prefetch,
            **kwargs,
        )

        validation_data_loader = None
        if validation_data is not None:
            validation_data_loader = ValidationDataLoader(
                dataset=validation_data,
                transform=transformation,
                batch_size=self.trainer.batch_size,
                ctx=self.trainer.ctx,
                dtype=self.dtype,
                num_workers=num_workers,
                num_prefetch=num_prefetch,
                **kwargs,
            )
```

@yx1215 Do you think you could look into this, maybe fix it using a PR?

**Comment by yx1215** [comment]
> So it seems there is in general some documentation missing regarding `shuffle_buffer_length` in `loader.py` and in `parallelized_loader.py`.
> 
> Additionally, I guess in most cases, the argument shuffle_buffer_length is implicitly passed as an argument. It should, however , be explicit just like `num_workers` or `num_prefetch`. Example below:
> 
> ```
>     def train(
>         self,
>         training_data: Dataset,
>         validation_data: Optional[Dataset] = None,
>         num_workers: Optional[int] = None,
>         num_prefetch: Optional[int] = None,
>         **kwargs,
>     ) -> Predictor:
>         has_negative_data = any(np.any(d["target"] < 0) for d in training_data)
>         low = -10.0 if has_negative_data else 0
>         high = 10.0
>         bin_centers = np.linspace(low, high, self.num_bins)
>         bin_edges = np.concatenate(
>             [[-1e20], (bin_centers[1:] + bin_centers[:-1]) / 2.0, [1e20]]
>         )
> 
>         logging.info(
>             f"using training windows of length = {self.train_window_length}"
>         )
> 
>         transformation = self.create_transformation(
>             bin_edges, pred_length=self.train_window_length
>         )
> 
>         training_data_loader = TrainDataLoader(
>             dataset=training_data,
>             transform=transformation,
>             batch_size=self.trainer.batch_size,
>             num_batches_per_epoch=self.trainer.num_batches_per_epoch,
>             ctx=self.trainer.ctx,
>             num_workers=num_workers,
>             num_prefetch=num_prefetch,
>             **kwargs,
>         )
> 
>         validation_data_loader = None
>         if validation_data is not None:
>             validation_data_loader = ValidationDataLoader(
>                 dataset=validation_data,
>                 transform=transformation,
>                 batch_size=self.trainer.batch_size,
>                 ctx=self.trainer.ctx,
>                 dtype=self.dtype,
>                 num_workers=num_workers,
>                 num_prefetch=num_prefetch,
>                 **kwargs,
>             )
> ```
> 
> @yx1215 Do you think you could look into this, maybe fix it using a PR?

Yes, I've created a new PR to add documentation as well as fix the problem of implicit input. Actually I've already add ```shuffle_buffer_length``` to the general ```GluonEstimator``` class in ```model/estimator.py```, so the problem only occurs when sub estimators reimplement the train method. Current I only find one model that does this which is the wavenet. Feel free to tell me if there is more. Here is the new PR #910 .
